[Dynamo] Use custom backend to reenter metadata tf mode when tracing while/cond

mlazos opened the pull request:
For tracing cond/while in eager, we trace the HOP with the eager backend with metadata torchfunction mode enabled. HOPs disallow the mutation that occurs in this torch function mode, so it is not able to be traced. As a result, we use a custom backend which enters this mode for tracing these HOPs. Thanks to @ydwu4 for the help with implementing this

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #135503
* #135502
* #135422
* #135444
* #135443
* #133137
* __->__ #134732



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

mlazos committed:
[Dynamo] Disable metadata tf mode when tracing cond

[ghstack-poisoned]

mlazos committed:
Update on "[Dynamo] Disable metadata tf mode when tracing cond"




[ghstack-poisoned]

mlazos committed:
Update on "[Dynamo] Disable metadata tf mode when tracing while/cond"

Previously, before dynamo handled torch function mode tracing, these modes were ignored. This keeps this behavior by popping this mode before tracing with dynamo in the handling of cond.




[ghstack-poisoned]

mlazos committed:
Update on "[Dynamo] Disable metadata tf mode when tracing while/cond"

Previously, before dynamo handled torch function mode tracing, these modes were ignored. This keeps this behavior by popping this mode before tracing with dynamo in the handling of cond.




[ghstack-poisoned]

mlazos committed:
Update on "[Dynamo] Disable metadata tf mode when tracing while/cond"

Previously, before dynamo handled torch function mode tracing, these modes were ignored. This keeps this behavior by popping this mode before tracing with dynamo in the handling of cond.




[ghstack-poisoned]

mlazos committed:
Update on "[Dynamo] Disable metadata tf mode when tracing while/cond"

Previously, before dynamo handled torch function mode tracing, these modes were ignored. This keeps this behavior by popping this mode before tracing with dynamo in the handling of cond.




[ghstack-poisoned]

mlazos committed:
Update on "[Dynamo] Disable metadata tf mode when tracing while/cond"

Previously, before dynamo handled torch function mode tracing, these modes were ignored. This keeps this behavior by popping this mode before tracing with dynamo in the handling of cond.




[ghstack-poisoned]

zou3519 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1217,6 +1222,11 @@ def __torch_function__(
 
 
+_temp_remove_metadata_torch_function_mode = _make_temp_remove_mode_context_manager(
+    TorchFunctionMetadataMode
```
@ydwu4 I'm confused because I thought we wanted to enable this mode inside the body of cond/while-loop?

ydwu4 replied:
Yeah, we should enable these otherwise the subgraphs will lose the metadata.

zou3519 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1245,6 +1255,11 @@ def __torch_function__(
         return func(*args, **kwargs)
 
 
+_temp_remove_pre_dispatch_torch_function_mode = _make_temp_remove_mode_context_manager(
+    PreDispatchTorchFunctionMode
+)
```
@mlazos do you have a test case that demonstrates what's going on?

mlazos replied:
Basically the reason I did this was there were some failures on the cond/while loop tests, I briefly discussed with @avikchaudhuri and we thought disabling this mode was fine since dynamo was previously not tracing it anyway, popping this preserves the existing behavior.

ydwu4 replied:
I think we should address the test failures. What tests are failing? I can help to see what happened.

mlazos replied:
Thanks Yidi, so the test failures are `test_cond_with_module_stack_export_with` on the next PR in this stack. Basically we graph break at line 1221 of proxy_tensor.py. If we're tracing is it strictly necessary that we keep track of those counts?

mlazos committed:
Update on "[Dynamo] Disable metadata tf mode when tracing while/cond"

Previously, before dynamo handled torch function mode tracing, these modes were ignored. This keeps this behavior by popping this mode before tracing with dynamo in the handling of cond.




[ghstack-poisoned]

mlazos committed:
Update on "[Dynamo] Disable metadata tf mode when tracing while/cond"

Previously, before dynamo handled torch function mode tracing, these modes were ignored. This keeps this behavior by popping this mode before tracing with dynamo in the handling of cond.




[ghstack-poisoned]

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -168,12 +170,15 @@ def _validate_input(pred, true_fn, false_fn, operands):
-                    pred, true_fn, false_fn, operands
-                )
+    with _set_compilation_env(), torch._dynamo.utils.disable_cache_limit(), _temp_remove_pre_dispatch_torch_function_mode():
+        with _temp_remove_metadata_torch_function_mode() as metadata_mode:
```
Do you mind also changing associative_scan's implementation? https://github.com/pytorch/pytorch/blob/main/torch/_higher_order_ops/associative_scan.py#L128-L132.

mlazos replied:
Yup I'll update

ydwu4 replied:
oops, I forgot flex attention: https://github.com/pytorch/pytorch/blob/main/torch/nn/attention/flex_attention.py#L1036-L1040

mlazos replied:
associative_scan doesn't use eager?

mlazos committed:
Update on "[Dynamo] Use custom backend to reenter metadata tf mode when tracing while/cond"

For tracing cond/while in eager, we trace the HOP with the eager backend with metadata torchfunction mode enabled. HOPs disallow the mutation that occurs in this torch function mode, so it is not able to be traced. As a result, we use a custom backend which enters this mode for tracing these HOPs. Thanks to ydwu4 for the help with implementing this




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

mlazos committed:
Update on "[Dynamo] Use custom backend to reenter metadata tf mode when tracing while/cond"

For tracing cond/while in eager, we trace the HOP with the eager backend with metadata torchfunction mode enabled. HOPs disallow the mutation that occurs in this torch function mode, so it is not able to be traced. As a result, we use a custom backend which enters this mode for tracing these HOPs. Thanks to ydwu4 for the help with implementing this




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

mlazos committed:
Update on "[Dynamo] Use custom backend to reenter metadata tf mode when tracing while/cond"

For tracing cond/while in eager, we trace the HOP with the eager backend with metadata torchfunction mode enabled. HOPs disallow the mutation that occurs in this torch function mode, so it is not able to be traced. As a result, we use a custom backend which enters this mode for tracing these HOPs. Thanks to ydwu4 for the help with implementing this




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

clee2000 commented:
@pytorchbot revert -m "broke functorch/test_control_flow.py::TestControlFlow::test_scan_simple_graph [GH job link](https://github.com/pytorch/pytorch/actions/runs/10804912306/job/29980571390) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/444b52ff40cf4afce7bc3fdcf021a88eab3b954c), newly added test yesterday" -c landrace

mlazos committed:
Update on "[Dynamo] Use custom backend to reenter metadata tf mode when tracing while/cond"

For tracing cond/while in eager, we trace the HOP with the eager backend with metadata torchfunction mode enabled. HOPs disallow the mutation that occurs in this torch function mode, so it is not able to be traced. As a result, we use a custom backend which enters this mode for tracing these HOPs. Thanks to ydwu4 for the help with implementing this




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

mlazos committed:
Update on "[Dynamo] Use custom backend to reenter metadata tf mode when tracing while/cond"

For tracing cond/while in eager, we trace the HOP with the eager backend with metadata torchfunction mode enabled. HOPs disallow the mutation that occurs in this torch function mode, so it is not able to be traced. As a result, we use a custom backend which enters this mode for tracing these HOPs. Thanks to ydwu4 for the help with implementing this




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

albanD commented:
@pytorchbot revert -m "Broke tests on main" -c forcemerged

See https://hud.pytorch.org/pytorch/pytorch/commit/c56728b643e2b7d796abd7ec45803319e1c5967d for example of failure

albanD commented:
@pytorchbot revert -m "Broke tests on main" -c ignoredsignal

Putting ignoredsignal since it was force merged.
See https://hud.pytorch.org/pytorch/pytorch/commit/c56728b643e2b7d796abd7ec45803319e1c5967d for the failure

mlazos committed:
Update on "[Dynamo] Use custom backend to reenter metadata tf mode when tracing while/cond"

For tracing cond/while in eager, we trace the HOP with the eager backend with metadata torchfunction mode enabled. HOPs disallow the mutation that occurs in this torch function mode, so it is not able to be traced. As a result, we use a custom backend which enters this mode for tracing these HOPs. Thanks to ydwu4 for the help with implementing this




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

mlazos commented:
@pytorchbot revert -m "broke python test/quantization/pt2e/test_numeric_debugger.py TestNumericDebugger.test_re_export_preserve_handle modified yesterday" -c landrace

mlazos committed:
Update on "[Dynamo] Use custom backend to reenter metadata tf mode when tracing while/cond"

For tracing cond/while in eager, we trace the HOP with the eager backend with metadata torchfunction mode enabled. HOPs disallow the mutation that occurs in this torch function mode, so it is not able to be traced. As a result, we use a custom backend which enters this mode for tracing these HOPs. Thanks to ydwu4 for the help with implementing this




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]
